runtimes/dart-2.18/function_types.dart

```diff
@@ -1,33 +1,77 @@
 import 'dart:convert';
 
 class Request {
-  final Map<String, dynamic> variables;
-  final Map<String, dynamic> headers;
-  final String payload;
-
-  Request({
-    this.variables = const {},
-    this.headers = const {},
-    this.payload = '',
-  });
+  String bodyString;
+  dynamic body;
+  Map<String, dynamic> headers;
+  String method;
+  String url;
+  String path;
+  int port;
+  String scheme;
+  String host;
+  String queryString;
+  Map<String, String> query;
+
+  Request({ String bodyString = '', dynamic body = '', Map<String, dynamic> headers = const {}, String method = '', String url = '', String path = '', int port = 80, String scheme = '', String host = '', String queryString = '', Map<String, String> query = const {} })
+    : bodyString = bodyString, body = body, headers = headers, method = method, url = url, path = path, port = port, scheme = scheme, host = host, queryString = queryString, query = query {
+  }
 }
 
 class Response {
-  int _status = 200;
-  dynamic _text;
+  dynamic send(String body, [int statusCode = 200, Map<String, dynamic> headers = const {}]) {
+    return {
+      'body': body,
+      'statusCode': statusCode,
+      'headers': headers,
+    };
+  }
+
+  dynamic json(Map<String, dynamic> json, [int statusCode = 200, Map<String, dynamic> headers = const {}]) {
+    var headersMerged = {
+      ...headers,
+      'content-type': 'application/json'
+    };
+    return this.send(jsonEncode(json), statusCode, headersMerged);
+  }
+
+  dynamic empty() {
+    return this.send('', 204, const {});
+  }
 
-  int get status => _status;
-  dynamic get body => _text;
+  dynamic redirect(String url, [int statusCode = 301, Map<String, dynamic> headers = const {}]) {
+    var headersMerged = {
+      ...headers,
+      'location': url
+    };
+    return this.send('', statusCode, headersMerged);
+  }
+}
+
+class Context {
+  Request req;
+  Response res;
+
+  List<String> logs = [];
+  List<String> errors = [];
+
+  Context(Request req, Response res)
+    : req = req, res = res {
+  }
 
-  Response send(String? text, {int status = 200}) {
-    _text = text;
-    _status = status;
-    return this;
+  void log(dynamic message) {
+    if (message is List || message is Map) { 
+      this.logs.add(jsonEncode(message));
+    } else {
+      this.logs.add(message.toString());
+    }
   }
 
-  Response json(Map<String, dynamic> json, {int status = 200}) {
-    _text = json;
-    _status = status;
-    return this;
+  void error(dynamic message) {
+    if (message is List || message is Map) { 
+      this.errors.add(jsonEncode(message));
```

Comment:
This could throw an exception if it's not possible to jsonEncode message.

Reply:
For example, if my function was this:

```dart
import 'dart:async';

class MyObject {
  String name;
  int age;
  MyObject({required this.name, required this.age});
}

Future<dynamic> main(final context) async {
  context.log([MyObject(name: 'name', age: 12)]);
  return context.res.json({
    'message': "Hello Open Runtimes 👋",
  });
}
```

An exception would be thrown and the `x-open-runtimes-errors` header would have:

```
Converting object to an encodable object failed: Instance of 'MyObject'
#0      _JsonStringifier.writeObject (dart:convert/json.dart:794:7)
#1      _JsonStringifier.writeList (dart:convert/json.dart:842:7)
#2      _JsonStringifier.writeJsonValue (dart:convert/json.dart:824:7)
#3      _JsonStringifier.writeObject (dart:convert/json.dart:785:9)
#4      _JsonStringStringifier.printOn (dart:convert/json.dart:983:17)
#5      _JsonStringStringifier.stringify (dart:convert/json.dart:968:5)
#6      JsonEncoder.convert (dart:convert/json.dart:345:30)
#7      JsonCodec.encode (dart:convert/json.dart:231:45)
#8      jsonEncode (dart:convert/json.dart:114:10)
#9      Context.log (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/function_types.dart:64:21)
#10     main (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/example/lib/main.dart:10:11)
#11     main.<anonymous closure>.<anonymous closure> (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/server.dart:107:36)
#12     _rootRun (dart:async/zone.dart:1391:13)
#13     _CustomZone.run (dart:async/zone.dart:1293:19)
#14     _runZoned (dart:async/zone.dart:1829:10)
#15     runZoned (dart:async/zone.dart:1751:10)
#16     main.<anonymous closure> (file:///Users/steven/Projects/open-runtimes-open-runtimes/runtimes/dart-2.18/server.dart:105:13)
<asynchronous suspension>
#17     handleRequest (package:shelf/shelf_io.dart:137:16)
<asynchronous suspension>
```

Reply:
Hmm what do you suggest? Wrap it into try/catch, and do toString as fallback?

Reply:
That sounds reasonable